pkg/typesystem/resolver.go

```diff
@@ -0,0 +1,102 @@
+package typesystem
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sync"
+
+	"github.com/oklog/ulid/v2"
+	"github.com/openfga/openfga/pkg/storage"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+)
+
+var (
+	modelSchemaVersionCounter = promauto.NewCounterVec(
+		prometheus.CounterOpts{
+			Name: "model_schema_version_counter",
+			Help: "A counter counting the different authorization model schema versions that have been resolved",
+		},
+		[]string{"schema_version"},
+	)
+)
+
+// TypesystemResolverFunc is a function that implementations can implement to provide lookup and
+// resolution of a Typesystem.
+type TypesystemResolverFunc func(ctx context.Context, storeID, modelID string) (*TypeSystem, error)
+
+// MemoizedTypesystemResolverFunc memoizes the provided TypesystemResolverFunc. It returns
+// the value from the first invocation of the underlying resolver function.
+//
+// The memoized resolver function is safe for concurrent use.
+func MemoizedTypesystemResolverFunc(fn TypesystemResolverFunc) TypesystemResolverFunc {
+	var cache sync.Map
```

Comment:
why do we need another cache of models? we already have one here https://github.com/openfga/openfga/blob/038c5c1f8c2dca1df38a32e04143fb162d476d83/pkg/storage/caching/caching.go#L22-L27

we can do the `modelSchemaVersionCounter++` in there

Reply:
The cached store caches the models, but it doesn't cache the resolution of the model and the typesystem constructed from it.

Instead of having to call `ReadAuthorizationModel` and `FindLatestAuthorizationModelID` everywhere we need it, we simply memoize the resolution of the typesystem once for each `storeID` and `modelID` we see. The memoization bypasses the cached storage layer altogether in favor of returning a prior constructed typesystem.

Reply:
Can't we change the existing cache to cache typesys instead of models?

Reply:
If we change our storage layer then yes. But I see that separate concerns.